src/riak_kv_vnode.erl

```diff
@@ -934,36 +958,15 @@ handle_coverage(?KV_LISTBUCKETS_REQ{item_filter=ItemFilter},
         _ ->
             {noreply, State}
     end;
-handle_coverage(#riak_kv_listkeys_req_v3{bucket=Bucket,
-                                         item_filter=ItemFilter},
-                FilterVNodes, Sender, State) ->
-    %% v3 == no backpressure
-    ResultFun = result_fun(Bucket, Sender),
-    Opts = [{bucket, Bucket}],
-    handle_coverage_keyfold(Bucket, ItemFilter, ResultFun,
-                            FilterVNodes, Sender, Opts, State);
-handle_coverage(?KV_LISTKEYS_REQ{bucket=Bucket,
-                                 item_filter=ItemFilter},
-                FilterVNodes, Sender, State) ->
-    %% v4 == ack-based backpressure
-    ResultFun = result_fun_ack(Bucket, Sender),
-    Opts = [{bucket, Bucket}],
-    handle_coverage_keyfold(Bucket, ItemFilter, ResultFun,
-                            FilterVNodes, Sender, Opts, State);
-handle_coverage(#riak_kv_index_req_v1{bucket=Bucket,
-                              item_filter=ItemFilter,
-                              qry=Query},
-                FilterVNodes, Sender, State) ->
-    %% v1 == no backpressure
-    handle_coverage_index(Bucket, ItemFilter, Query,
-                          FilterVNodes, Sender, State, fun result_fun/2);
-handle_coverage(?KV_INDEX_REQ{bucket=Bucket,
-                              item_filter=ItemFilter,
-                              qry=Query},
-                FilterVNodes, Sender, State) ->
-    %% v2 = ack-based backpressure
-    handle_coverage_index(Bucket, ItemFilter, Query,
-                          FilterVNodes, Sender, State, fun result_fun_ack/2).
+handle_coverage_request(kv_index_request, Req, FilterVNodes, Sender, State) ->
+    Bucket = riak_kv_requests:get_bucket(Req),
+    ItemFilter = riak_kv_requests:get_item_filter(Req),
+    Query = riak_kv_requests:get_query(Req),
+    ResultFun = case riak_kv_requests:get_ack_backpressure(Req) of
+                    true -> fun result_fun_ack/2;
+                    false -> fun result_fun/2
+                end,
+    handle_coverage_index(Bucket, ItemFilter, Query, FilterVNodes, Sender, State, ResultFun).
 
 -spec prepare_index_query(?KV_INDEX_Q{}) -> ?KV_INDEX_Q{}.
 prepare_index_query(#riak_kv_index_v3{term_regex=RE} = Q) when
```

Comment:
Further down we have:
```
 %% @HACK
             %% Really this should be decided in the backend
             %% if there was a index_query fun.
             FoldType = case riak_index:return_body(Query) of
                            true -> fold_objects;
                            false -> fold_keys
                        end,
```

This should be put into a function `fold_type_for_query/1`